### ISSUE TITLE
Refactoring plotting

### ISSUE THREAD

**Issue by trappmartin** [TuringLang/MCMCChain.jl#15]
The current implementation has the following drawbacks:

- does not use User Recipes
- the current implementation lacks diagnostics plots #8 
- the size of the plot is not dynamically adjusted depending on the number of parameters

**Comment by trappmartin** [comment]
@hessammehr I just tried your branch, and it seems that the plotting of multiple plotting types works just fine!

I used:
```julia
plot(chn, [MCMCChain.TracePlot, MCMCChain.DensityPlot])
```
so I guess we can simply use your implementation instead and export the user recipe types.
What do you think?

**Comment by hessammehr** [comment]
@trappmartin Nice, are the size issued resolved on the GR backend? 

Two things to add:

- `@userplot` automagically exports the various plotting functions, e.g. `traceplot`, `densityplot`, so perhaps we need to think about whether we really need to export the plot types as well. Interested to know your opinion.
- I think we need a default `plot(::Chains)` function for with sensible defaults. PyMC3 has `chain.plot()` for instance and that almost always does the right thing. A type recipe would seem appropriate for that and I can have a go at it tonight/tomorrow. What do you think?

**Comment by hessammehr** [comment]
The weird sizing issues are fixed now because we manually calculate the required size for the overall plot. Things look pretty good using the PyPlot backend but fonts look funny on GR and tweaking the `fontfamily` doesn't seem to change anything.

@trappmartin  do you think we could merge this in the mean time?

**Comment by hessammehr** [comment]
@trappmartin I was just looking at a set of plots generated by PyMC3. Random vectors are visualized very nicely (different colours instead of different subplots), and the default plotting behaviour might be something that we might want to aim for, i.e. trace and density plots (for continuous variables at least).

Interested to know your thoughts on this.

![trace](https://user-images.githubusercontent.com/392251/45374409-e93e9400-b5e9-11e8-9df9-276a33af923d.PNG)

**Comment by trappmartin** [comment]
this is a very good idea. Currently, the colours indicate the run. Do you know how this is handled in PyMC3? I'll have a closer look once the compiler refactoring is done.

**Comment by trappmartin** [comment]
How about we have a colordim kwarg that allows us to specify how the colouring & stacking should be applied.
E.g.
```julia
julia> plot(chn::AbstractChain; colordim = :parameter) # color paramters, stack chains
julia> plot(chn::AbstractChain; colordim = :chain) # color chains, stack paramters
julia> traceplot(chn::AbstractChain, i::Int; colordim = :parameter) # color paramters, i selects chain
julia> traceplot(chn::AbstractChain, i::Int; colordim = :chain) # color chains, i selects parameter
```

**Comment by trappmartin** [comment]
Closing in favour of PR #17
